src/chainstate/stacks/miner.rs

```diff
@@ -326,6 +393,218 @@ impl StacksBlockBuilder {
 
         test_debug!("\n\nMiner {}: Finished mining child of {}/{}. Trie is in mined_blocks table.\n", self.miner_id, self.chain_tip.burn_header_hash, self.chain_tip.anchored_header.block_hash());
     }
+    
+    /// Unconditionally build an anchored block from a list of transactions.
+    /// Used when we are re-building a valid block after we exceed budget
+    fn make_anchored_block_from_txs(mut builder: StacksBlockBuilder, chainstate: &mut StacksChainState, mut txs: Vec<StacksTransaction>) -> Result<StacksBlock, Error> {
+        debug!("Build anchored block from {} transactions", txs.len());
+        let mut epoch_tx = builder.epoch_begin(chainstate)?;
+        for tx in txs.drain(..) {
+            builder.try_mine_tx(&mut epoch_tx, &tx)?;
+        }
+        let block = builder.mine_anchored_block(&mut epoch_tx);
+        builder.epoch_finish(epoch_tx);
+        Ok(block)
+    }
+
+    /// Create a block builder for mining
+    fn make_block_builder(burndb: &mut BurnDB,
+                          chainstate: &mut StacksChainState,
+                          parent_burn_header_hash: &BurnchainHeaderHash,
+                          parent_block_hash: &BlockHeaderHash,
+                          proof: VRFProof,
+                          pubkey_hash: Hash160) -> Result<StacksBlockBuilder, Error> {
+
+        let builder = 
+            if *parent_burn_header_hash == FIRST_BURNCHAIN_BLOCK_HASH && *parent_block_hash == FIRST_STACKS_BLOCK_HASH {
+                StacksBlockBuilder::first_pubkey_hash(0, parent_burn_header_hash, FIRST_BURNCHAIN_BLOCK_TIMESTAMP, &proof, pubkey_hash)
+            }
+            else {
+                // building off an existing stacks block
+                let parent_stacks_block_snapshot = {
+                    let mut tx = burndb.tx_begin()?;
+                    let parent_stacks_block_snapshot = BurnDB::get_block_snapshot_for_winning_stacks_block(&mut tx, parent_burn_header_hash, parent_block_hash)?
+                        .ok_or(Error::NoSuchBlockError)?;
+
+                    parent_stacks_block_snapshot
+                };
+
+                let parent_chain_tip = StacksChainState::get_anchored_block_header_info(&chainstate.headers_db, parent_burn_header_hash, parent_block_hash)?
+                    .ok_or(Error::NoSuchBlockError)?;
+
+                let new_work = StacksWorkScore {
+                    burn: parent_stacks_block_snapshot.total_burn,
+                    work: parent_chain_tip.block_height.checked_add(1).expect("FATAL: block height overflow")
+                };
+
+                StacksBlockBuilder::from_parent_pubkey_hash(0, &parent_chain_tip, &new_work, &proof, pubkey_hash)
+            };
+
+        Ok(builder)
+    }
+
+    /// Given access to the mempool, mine an anchored block with no more than the given execution cost.
+    pub fn build_anchored_block(burndb: &mut BurnDB,
```

Comment:
If my understanding is correct -- during normal operation, any mempool transaction that gets mined would remain in the mempool db until it gets garbage collected (whether it was mined or not). The thing that prevents the transaction from being mined again is that process_transaction will fail because of a nonce check failure, right? Is this behavior we want to keep going forward, or should we start an issue for a more proactive dropping (or removing from consideration while a reorg hasn't occurred) of mined transactions?

Reply:
> The thing that prevents the transaction from being mined again is that process_transaction will fail because of a nonce check failure, right?

Yes, that's correct.

> Is this behavior we want to keep going forward, or should we start an issue for a more proactive dropping (or removing from consideration while a reorg hasn't occurred) of mined transactions?

I don't think we should ever drop transactions preemptively unless they can never be valid on any fork.  I think instead the miner should maintain some hints on a per-fork basis as to which transactions can be skipped over when building on a descendant chain tip that can never include them.

Reply:
Yep, that works for me -- we can revisit per-fork hints if this turns out to be a problem